README.rst
Python wrapper for the 3di subgrid library
==========================================

Prerequisites
-----

We need the compiled Fortran subgrid library. There are a couple of common
locations where we look for it.::

   .
   ~/local/lib
   ~/.local/lib
   /opt/3di/lib
   /usr/local/lib
   /usr/lib

A convention on linux is to install the library into ``/opt/3di/``.

In case you have an alternative location, you can set the ``SUBGRID_PATH``
environment variable::

   $ export SUBGRID_PATH=/home/user/svn/3di/trunk/subgridf90/src/.libs

(On windows the command is ``set`` instead of ``export``).


Ubuntu install of the Fortran library
-------------------------------------

If the most recently released version is good enough, you can use the
`ubuntu package instructions
<http://jenkins.3di.lizard.net/ubuntu/precise64/AN_INTRODUCTION_ON_HOW_TO_USE_THIS.html>`_
to get the latest version of the subgrid library.


Mostly-standard compilation of the Fortran library on Ubuntu
------------------------------------------------------------

For the python wrapper we of course need the actual subgrid Fortran
library so that we can wrap it. A mostly standard installation on
ubuntu (which is also used on the jenkins test server and on the demo
website) uses the standard ubuntu netcdf packages and fortran::

    $ sudo apt-get install libnetcdf-dev \
      libnetcdf6 \
      gfortran \
      libshp-dev \
      libgdal1-dev

For the python netcdf library compilation we need another package::

    $ sudo apt-get install libhdf5-serial-dev

We also need the fortrangis library. We can compile it by hand (and
add a couple of dependencies), but using the "ubuntu install"
mentioned above is handier. Just do an ``apt-get install fortrangis``
after you set up the ubuntu repository.

And we use the regular configure/make/make install steps with two changes::

    $ ./autogen.sh  # Try it without. If something breaks, run it.
    $ FCFLAGS="-I/usr/include -g -O0" ./configure --prefix=/opt/3di --with-gdal
    $ make
    $ make install

The two changes:

- ``FCFLAGS`` to let fortran find the ubuntu netcdf packages.

- A ``--prefix`` to install it into ``/opt/3di/``.

Setup
------
Before you can use a python package you have to add it to your path. The python-subgrid package works in buildout (under linux)
and can also run in a virtualenv. The buildout way::

  python ./bootstrap.py # optionally with -v 1.0.0
  bin/buildout

The virtualenv way (assumes virtualenvwrapper and virtualenv are installed)::

  mkvirtualenv main
  workon main
  pip install ./python-subgrid
  # or if you want to add your source directory to the path
  python setup.py develop

Combination Windows and Anaconda::

- Download + Install Anaconda
- Download and install the :faulthandler: package, from http://www.lfd.uci.edu/~gohlke/pythonlibs/#faulthandler
- Download and install the :NetCDF4: package, from http://www.lfd.uci.edu/~gohlke/pythonlibs/#netcdf4
- Open an (Anaconda) Python-terminal (Press :[Ctrl]:+:[Alt]+:A:).
- pip.bat install -e ``<path to your python-subgrid GIT working copy>``


Usage
-----

There are two ways to use the wrapper. A handy way is as a context
manager, so with a ``with`` statement::

    with SubgridWrapper(mdu='/full/path/model.mdu') as subgrid:
        # subgrid is the actual fortran library.
        subgrid.something()

The second way is by calling :meth:`start` and :meth:`stop` yourself and
using the :attr:`library` attribute to access the Fortran library::

    wrapper = SubgridWrapper(mdu='/full/path/model.mdu')
    wrapper.start()
    wrapper.library.something()
    ...
    wrapper.stop()

Note: Without the ``mdu`` argument, no model is loaded and you're free to
use the library as you want.


Automatic tests
---------------

The code is tested on http://jenkins.3di.lizard.net. The ``libsubgrid.so``
that is used there is the one from the last successful build of
``subgridf90``, which is placed in ``/opt/3di/``.

The functional tests need testcases. Run ``update_testcases.sh`` to check them
out and test them. The tests find them automatically, so a simple ``bin/test``
is enough.

You can symlink those directories next to this README. You can also set
the ``SCENARIO_BASEDIR`` environment variable. Either set it globally or run
the tests like this, for instance::

    $ SCENARIO_BASEDIR=/some/directory bin/test

If the scenarios cannot be found, the functional model tests are skipped, btw.


Convenience scripts
-------------------

The python subgrid library contains some scripts that can be used to update input files::

  update-subgrid-network
  update-subgrid-tables
  convert_subgrid_mdu

The script ``update-subgrid-network`` updates input files from the old format (2x -1 in ``network.inp``)  to the new format.
The script ``update-subgrid-tables`` generates the ``*.tbl`` and ``*.grd`` files to the current format. These files can be used to speed-up initialisation.
The script ``convert_subgrid_mdu`` converts an old MDU file (version < 2) to a new MDU format version.

For details on the usage of these scripts please see::

  update-subgrid-network --help
  update-subgrid-tables --help
  convert_subgrid_mdu --help


CHANGES.rst
Changelog of python-subgrid
===================================================


0.25 (unreleased)
-----------------

- Nothing changed yet.


0.24 (2018-05-14)
-----------------

- Explicitly set typemap char value to 'S1'.


0.23 (2015-07-09)
-----------------

- Raingrid now fetches rain from raster-server and interpolates not anymore.


0.21 (2015-03-12)
-----------------

- Expose friction parameters (fricttype, const_friction, and frictcoefuser) corresponds with model reversion 1738.

- Default slice for velocities on interfaces (u1) and expose number of u, v velocity points (liutot, livtot).


0.20 (2014-12-04)
-----------------

- Made subgridpy prettier.

- Added option Outputdir (use with subgrid 0.7-5 or higher).

- Updated tests.


0.19 (2014-08-27)
-----------------

- Added wrapper.update_tables function. You need to call this function
  after each grid edit action.


0.18 (2014-08-06)
-----------------

Note: this upgrade needs subgrid version 0.7-0 or higher.

Note: you need to install Six system-wide when running buildout or scikit-image
will not build. `sudo easy_install -U six`

- Added draw_shape_on_raster in plotting.

- Point repos in update_testcases.sh to production model databank
  (hg.lizard.net).

- Added features to subgridpy, the offline python subgrid runner:

  - Specify a scenario dir and it tries to read area_wide_rain_grids.json and
    radar_grids.json

  - Area wide rain implements the "ontwerpbuien" from 3Di Live.

  - Radar rain implements the historic rain from 3Di Live.

  - Specify a radar rain using the command line.

  - Specify a area wide rain using the command line.

  - All (important parts) unit tested!


0.17 (2014-05-21)
-----------------

Note: You need subgrid version 0.6-1 (2014-05-14) or newer for the logging to work.

- Renamed set_mh_c_callback to set_logger.

- Added test test_001_run_hhnk.

- Added test test_001_run_delfland_default.

- Added test test_001_load_duifpolder_default_raingrid.

- Added test for duifp_default.

- More subgridpy options.

- set_logger is now optional in wrapper.


0.16 (2014-03-31)
-----------------

- Added uc, zg to SLICES.

- Added particle module + example (dependencies disabled for now)


0.15 (2014-03-24)
-----------------

- Nothing changed yet.


0.14 (2014-03-18)
-----------------

- get_nd now returns strings (wkt) as is (as an array). You have to convert it
  yourself.


0.13 (2014-03-18)
-----------------

- Fixed: visualization with 1d+2d models.

- Add crop/soil/infiltration/interception

- Cleanup tests

- Added restart option.


0.12 (2014-02-26)
-----------------

- Nothing changed yet.


0.11 (2014-02-19)
-----------------

- Test.


0.10 (2014-02-19)
-----------------

- Added extra check for 'id' before str.rstripping.


0.9 (2014-02-19)
----------------

- Nothing changed yet.


0.8 (2014-02-05)
----------------

- More tests and fixes for orifices.


0.7 (2014-01-30)
----------------

- Added RainGrid functionality with tests.


0.6 (2014-01-09)
----------------

- Updated for subgrid 0.3.


0.5 (2013-12-16)
----------------

- Nothing changed yet.


0.4 (2013-11-20)
----------------

- Nothing changed yet.


0.3 (2013-11-14)
----------------

- Added ``update_testcases.sh`` script for checking out the testcases that are
  needed for our functional tests.

- Started documentation on Fortran variables that you can call
  directly. All variables are documented (as undocumented variables
  raise a ``NotDocumentedError`` exception).

- Added a status page to the documentation with the svn and github
  locations and so on.

- Updated the build instructions, including hint to use the
  now-available ubuntu packages.

- Fixed library search order: specific wins over generic
  (=``/usr/lib``).

- Moved to github. Github pull requests are quite essential now for
  proper development. https://github.com/nens/python-subgrid . Mail
  Reinout for access if needed.

- Added roadmap document (in the sphinx docs in ``doc/``) that
  describes the main structure and future roadmap of this library.


0.2 (2013-09-23)
----------------

- Made a branch off an older stable version to create a 0.2 release.
  This is the "svn revision 714" version that was/is used on the server.

- The fortran library can be loaded through a "with" context manager now. The
  previous version loaded it upon importing the ``wrapper.py`` file, handling
  everything on the main module level.

  The current approach is cleaner and takes care of startup/shutdown code and
  model initialization/cleanup automatically. The latter depends on the
  fortran cleanup code to function well, which at the moment is not the case
  yet.

- The context manager behaviour is now also available with simple
  ``start()``/``stop()`` methods so that it can be used on the webserver where
  there's no single block-within-a-``with``-statement.

- Big documentation update. Sphinx documentation added (currently
  automatically rendered to http://jenkins.3di.lizard.net/doc/). Docstrings
  everywhere.


0.1 (2013-06-04)
----------------

- Refactored the wrapper to make the code cleaner and more testable.

- If the models aren't available, the functional model tests are skipped. This
  makes for quicker tests if you want to test just the internal unittests.

- Modified library loading routine to automatically look in a couple of
  standard locations, amongst them ``/opt/3di/``.

- Added code from the previous ``python_wrapper`` directory.

- Initial project structure created with nensskel 1.33.dev0.


notebooks/alcatraz2.py
import logging
import datetime
import functools
import itertools

# enum34
import enum

import pandas
import netCDF4
import numpy as np
import osgeo.osr
import dateutil.rrule

import tables
import docopt
import pyprind
import python_subgrid.particles


# Reason for termination
class Reason(enum.Enum):
    # VTK termination reasons
    OUT_OF_DOMAIN = 1
    NOT_INITIALIZED = 2
    UNEXPECTED_VALUE = 3
    OUT_OF_TIME = 4
    OUT_OF_STEPS = 5
    STAGNATION = 6
    # this is extra for living particles
    DEATH = 11
    ESCAPE = 12


def behaviour(txyz, velocity, targets, survivaltime=60*60*5, i0=0, mindistance=150):
    """swim with velocity to a target in the target list"""
    # determine target at t0
    x, y = txyz[0, 1], txyz[0, 2]
    dt = txyz[-1, 0] - txyz[0, 0]

    distance = np.sqrt((targets['x_utm'] - x)**2 + (targets['y_utm'] - y)**2)
    name = distance.argmin()
    target = targets.ix[name]

    # Default reason
    reason = Reason.OUT_OF_TIME.value

    # No land in sight (foggy night)
    if distance[name] > mindistance:
        return txyz, reason

    # Make sure we don't overshoot
    if distance.ix[name]/dt <= velocity:
        velocity = min(distance.ix[name]/dt, velocity)
        reason = Reason.ESCAPE.value

    angle = np.arctan2(target['y_utm'] - y, target['x_utm'] - x)
    dx = np.cos(angle) * velocity
    dy = np.sin(angle) * velocity
    # add the velocity to the path
    txyz[:, 1] += dx * txyz[:, 0]
    txyz[:, 2] += dy * txyz[:, 0]

    if (txyz[-1, 0] - (i0 * 900)) > survivaltime:
        reason = Reason.DEATH.value
    return txyz, reason


def num2deg(xtile, ytile, zoom):
    """convert x,y zoom number to a lat/long"""
    n = 2.0 ** zoom
    lon_deg = xtile / n * 360.0 - 180.0
    lat_rad = np.arctan(np.sinh(np.pi * (1 - 2 * ytile / n)))
    lat_deg = np.degrees(lat_rad)
    return (lon_deg, lat_deg)


def seed_alcatraz(system, velocity=None, behaviour=None):
    """toss in some swimmers"""
    n_new = 50
    x_src = np.random.uniform(locations.ix['alcatraz']['x_utm'], locations.ix['alcatraz']['x_utm']+100,
                              size=n_new)
    y_src = np.random.uniform(locations.ix['alcatraz']['y_utm'], locations.ix['alcatraz']['y_utm']+100,
                              size=n_new)
    system.reseed(
        pts=np.c_[x_src, y_src, np.zeros_like(x_src)],
        velocity=velocity,
        behaviour=behaviour
    )


def seed_domain(system, n_total=2000):
    n_current = len(system.get_ids())
    if n_current < n_total:
        n_new = n_total - n_current
    else:
        n_new = 0
    x_src = np.random.uniform(ll_utm[0], ur_utm[0], size=n_new)
    y_src = np.random.uniform(ll_utm[1], ur_utm[1], size=n_new)
    system.reseed(pts=np.c_[x_src, y_src, np.zeros_like(x_src)])


if __name__ == "__main__":
    doc = """
    Alcatraz Escape

    Usage:
        alcatraz2.py [--seed]
        alcatraz2.py behaviour <speed> [--visibility=<m>] [--seed]
        alcatraz2.py velocity <u> <v> [--seed]
        alcatraz2.py --version

    Options:
        -h --help     Show this screen.
        --version     Show version.
        --visibility=<m>  Visibility [default: 800].
        --seed        Seed the domain with extra particles
    """
    arguments = docopt.docopt(doc, version='Alcatraz Escape 1.0')
    print(arguments)

    visibility = float(arguments['--visibility'])

    use_velocity = arguments['velocity']
    u = arguments['<u>']
    v = arguments['<v>']
    if use_velocity:
        u = float(u)
        v = float(v)

    use_behaviour = arguments['behaviour']
    speed = arguments['<speed>']
    if use_behaviour:
        speed = float(speed)

    seed = arguments['--seed']

    logging.basicConfig()
    logging.root.setLevel(logging.WARNING)

    # generate the spatial reference systems and the transformations
    utm_srs = osgeo.osr.SpatialReference()
    utm_srs.ImportFromEPSGA(26910)
    osm_srs = osgeo.osr.SpatialReference()
    osm_srs.ImportFromEPSGA(3857)
    wgs_srs = osgeo.osr.SpatialReference()
    wgs_srs.ImportFromEPSGA(4326)
    utm2wgs = osgeo.osr.CoordinateTransformation(utm_srs, wgs_srs)
    utm2osm = osgeo.osr.CoordinateTransformation(utm_srs, osm_srs)
    wgs2utm = osgeo.osr.CoordinateTransformation(wgs_srs, utm_srs)
    wgs2osm = osgeo.osr.CoordinateTransformation(wgs_srs, osm_srs)

    # define the location of the Alcatraz escape location
    locations = []
    locations.append(dict(name='alcatraz', lon=-122.4239052, lat=37.8279125))
    # locations.append(dict(name='point_blunt', lat=37.852489, lon=-122.419121))
    locations.append(dict(name='golden_gate', lat=37.831936, lon=-122.472733))
    # locations.append(dict(name='fishermans_warf', lat=37.810781, lon=-122.412790))
    locations.append(dict(name='a', lat=37.832079, lon=-122.472916))
    locations.append(dict(name='b', lat=37.825529, lon=-122.479080))
    locations.append(dict(name='c', lat=37.825900, lon=-122.482869))
    locations.append(dict(name='d', lat=37.826160, lon=-122.486745))
    locations.append(dict(name='e', lat=37.822314, lon=-122.495721))
    locations.append(dict(name='f', lat=37.819975, lon=-122.499629))
    locations.append(dict(name='g', lat=37.820970, lon=-122.504606))
    locations.append(dict(name='h', lat=37.815442, lon=-122.528458))
    locations = pandas.DataFrame(locations)

    xyz = wgs2osm.TransformPoints(
        list(np.c_[locations['lon'], locations['lat']])
    )
    locations['x_osm'] = [xyz_i[0] for xyz_i in xyz]
    locations['y_osm'] = [xyz_i[1] for xyz_i in xyz]

    xyz = wgs2utm.TransformPoints(
        np.c_[locations['lon'], locations['lat']]
    )
    locations['x_utm'] = [xyz_i[0] for xyz_i in xyz]
    locations['y_utm'] = [xyz_i[1] for xyz_i in xyz]
    locations = locations.set_index('name')


    # we're using this area
    ll, ur = num2deg(1308, 3167, 13), num2deg(1314, 3163, 13)
    # lookup bounding box in different coordinate systems
    ll_osm = wgs2osm.TransformPoint(ll[0], ll[1])
    ur_osm = wgs2osm.TransformPoint(ur[0], ur[1])

    ll_utm = wgs2utm.TransformPoint(ll[0], ll[1])
    ur_utm = wgs2utm.TransformPoint(ur[0], ur[1])

    ll, ur, ll_osm, ur_osm, ll_utm, ur_utm

    # Model input
    filename = '/Users/baart_f/models/sfo/sfo-3di/subgrid_map_15min.nc'
    ds = netCDF4.Dataset(filename)
    times = netCDF4.num2date(ds.variables['time'][:],
                             ds.variables['time'].units)
    system = python_subgrid.particles.ParticleSystem(ds)

    dfs = []
    selection = times[250:340]
    # 1) initialization with number of iterations
    prbar = pyprind.ProgBar(selection.shape[0],
                            width=100, track_time=True)

    rrule = dateutil.rrule.rrule(
        dateutil.rrule.MINUTELY,
        dtstart=datetime.datetime(1962, 6, 11, 19, 0),
        interval=30
    )
    swimmers = rrule.between(
        datetime.datetime(1962, 6, 11, 19, 0),
        datetime.datetime(1962, 6, 12, 4, 0),
        inc=True
    )
    targets = locations[locations.index != 'alcatraz']

    behaviour = functools.partial(behaviour,
                                  velocity=speed,
                                  targets=targets,
                                  mindistance=visibility)
    if use_velocity:
        velocity = (u, v)
    else:
        velocity = None
    for i, t in enumerate(selection):
        prbar.update()
        if t in swimmers:
            if use_behaviour:
                behaviour = functools.partial(behaviour, i0=i)
            else:
                behaviour = None
            seed_alcatraz(system, behaviour=behaviour, velocity=velocity)
        else:
            if seed:
                seed_domain(system)
        system.update_grid(i=i)
        system.tracer.update()
        df = system.get_particles()
        dfs.append(df)
    df = pandas.concat(dfs)
    df.drop_duplicates(cols=['particle', 't'], inplace=True)
    df.sort(columns=['particle', 't'], inplace=True)
    # Add the particles to the file
    filename = 'parts.h5'

    # Create a new group
    f = tables.open_file(filename, 'a')
    for i in itertools.count():
        name = 'particles_%d' % (i,)
        if name in f.root._v_children:
            continue
        else:
            break
    f.close()
    # Store the table
    df.to_hdf(filename, name)
    # Add the parameters
    f = tables.open_file(filename, 'a')
    f.set_node_attr("/" + name, 'visibility', visibility)
    f.set_node_attr("/" + name, 'speed', speed)
    f.set_node_attr("/" + name, 'u', u)
    f.set_node_attr("/" + name, 'v', v)
    f.set_node_attr("/" + name, 'velocity', velocity)
    f.set_node_attr("/" + name, 'use_behaviour', use_behaviour)
    f.set_node_attr("/" + name, 'use_velocity', use_velocity)
    f.set_node_attr("/" + name, 'seed', seed)
    f.close()
